### ISSUE TITLE
Live-ISO test failing

### ISSUE THREAD

**Issue by lentzi90** [metal3-io/baremetal-operator#2110]
### Which jobs are failing?

Periodic E2E: https://github.com/metal3-io/baremetal-operator/actions/workflows/e2e-test-periodic-main.yml

### Which tests are failing?

live-iso: https://github.com/metal3-io/baremetal-operator/actions/runs/12227994617/job/34105542268

### Since when has it been failing?

Fist failure was 4th December

### Reason for failure (if possible)

The BMH cannot be provisioned because the "image is not valid for use".

From Ironic logs:
```
2024-12-05 14:23:05.425 1 DEBUG ironic.common.images [None req-f1216348-13a7-4a46-bfcb-8de7d88b3fb1 - - - - - -] Image http://192.168.222.1/sysrescue-out.iso downloaded in 2.58 seconds. fetch_into /usr/lib/python3.9/site-packages/ironic/common/images.py:386
2024-12-05 14:23:05.430 1 DEBUG oslo_utils.imageutils.format_inspector [None req-f1216348-13a7-4a46-bfcb-8de7d88b3fb1 - - - - - -] Format inspector for vmdk does not match, excluding from consideration (Signature KDMV not found: b'3\xed\x90\x90') _process_chunk /usr/lib/python3.9/site-packages/oslo_utils/imageutils/format_inspector.py:1365
2024-12-05 14:23:05.435 1 DEBUG oslo_utils.imageutils.format_inspector [None req-f1216348-13a7-4a46-bfcb-8de7d88b3fb1 - - - - - -] Format inspector for vhdx does not match, excluding from consideration (Region signature not found at 30000) _process_chunk /usr/lib/python3.9/site-packages/oslo_utils/imageutils/format_inspector.py:1365
2024-12-05 14:23:05.437 1 ERROR ironic.common.images [None req-f1216348-13a7-4a46-bfcb-8de7d88b3fb1 - - - - - -] Security: The requested user image for the deployment node image cache failed to be able to be parsed by the image format checker: Multiple formats detected: iso,gpt: oslo_utils.imageutils.format_inspector.ImageFormatError: Multiple formats detected: iso,gpt
2024-12-05 14:23:05.512 1 ERROR ironic.conductor.utils [None req-f1216348-13a7-4a46-bfcb-8de7d88b3fb1 - - - - - -] Node f9ba65c4-987a-42b7-8965-1787c849a3f5 failed deploy step {'step': 'deploy', 'priority': 100, 'argsinfo': None, 'interface': 'deploy'}: The requested image is not valid for use.: ironic.common.exception.InvalidImage: The requested image is not valid for use.
```

### Anything else we need to know?

We thought first that the issue was that we did not specify the image hash, but this has been ruled out in https://github.com/metal3-io/baremetal-operator/pull/2103.

### Label(s) to be applied

/kind failing-test
One or more /area label. See https://github.com/metal3-io/baremetal-operator/labels for the list of labels.

**Comment by tuminoid** [comment]
Added to CI tracker.

**Comment by lentzi90** [comment]
I wonder if it could be related to this: https://bugs.launchpad.net/nova/+bug/2091114
We did a [temporary workaround](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/2302/commits/0a28aaeece6005752c59cfcfe429da819d120652) for it in CAPO.

**Comment by Rozzii** [comment]
I was expecting this originally: https://opendev.org/openstack/ironic/commit/669304bc0c6b2762c872b71297480c4b4ffdb554

**Comment by Rozzii** [comment]
Based on the quay artifacts, some change has happened between Nov 6 and Dec 3.

**Comment by lentzi90** [comment]
I think you are correct that this came with oslo.utils.
The specific commit that introduced it in oslo.utils is here: https://github.com/openstack/oslo.utils/commit/91af49beed6eea273863094bd62dcfb6688bfb11
Then there was some related changes that looks like they would be relevant to the live-iso use-case: https://github.com/openstack/oslo.utils/commit/3d4ae167cc1cc3509669a2c30f44e8d31ba39e3f
I haven't figured out yet how we are supposed to "allow" multiple formats.

**Comment by Rozzii** [comment]
I think in the test system we supposed to have the safety check diasbled as we don't really care, but I have thought that enabling all the formats is the default.

**Comment by tuminoid** [comment]
BMO 0.8 virtualmedia works, where ironic is pinned to 26.0: https://github.com/metal3-io/baremetal-operator/pull/2111

**Comment by tuminoid** [comment]
This is marked for 0.9. We have 2 rounds of workarounds implemented now, but the actual fix is pending. Should we close this as implemented in 0.9 and create another for 0.10 for the proper fix?

**Comment by lentzi90** [comment]
Sure we can do that or just change the milestone. Doesn't matter to me

**Comment by tuminoid** [comment]
It seems the revert is still failing, meaning main/0.9 is still broken. I guess whatever fix is going to be needs to be in 0.9 later on so let's actually keep 0.9 here.

**Comment by Rozzii** [comment]
Eventually in the Ironic community we have landed on this fix:
https://review.opendev.org/c/openstack/ironic/+/938363

**Comment by iurygregory** [comment]
Since the workaround in ironic is merged, should we https://review.opendev.org/c/openstack/ironic/+/938363, should we un-pin the ironic image? we need for https://github.com/metal3-io/baremetal-operator/pull/2229

**Comment by lentzi90** [comment]
Yes we should, but the tests are failing on the [revert](https://github.com/metal3-io/baremetal-operator/pull/2113). I am debugging it at the moment

**Comment by dtantsur** [comment]
@lentzi90 it's not the same test that is failing, correct? I see the live ISO one passing, what fails is `[It] provisions a BMH, applies detached and status annotations, then deprovisions`.

**Comment by dtantsur** [comment]
```
2025-02-24 08:38:45.115 1 DEBUG ironic.drivers.modules.agent_base [-] deploy command status for node 155e7ecd-923d-4355-9456-2a398fddc33f on step {'step': 'write_image', 'priority': 80, 'argsinfo': None, 'interface': 'deploy'}: {'id': 'c6b68d87-b5cf-431e-bfe2-661eeb74b45f', 'command_name': 'execute_deploy_step', 'command_status': 'FAILED', 'command_error': {'type': 'CommandExecutionError', 'code': 500, 'message': 'Command execution failed', 'details': 'Unexpected error while running command.\nCommand: /opt/ironic-python-agent/bin/python3 -m oslo_concurrency.prlimit --as=2147483648 -- [\'env\', \'LC_ALL=C\', \'LANG=C\', \'qemu-img\', \'info\', \'/tmp/cirros-0.6.2-x86_64-disk.img\', \'--output=json\']\nExit code: 1\nStdout: \'\'\nStderr: "python3 -m oslo_concurrency.prlimit: failed to execute [\'env\', \'LC_ALL=C\', \'LANG=C\', \'qemu-img\', \'info\', \'/tmp/cirros-0.6.2-x86_64-disk.img\', \'--output=json\']: [Errno 2] No such file or directory\\n"'}, 'command_result': None} process_next_step /usr/lib/python3.9/site-packages/ironic/drivers/modules/agent_base.py:1094
```
This is Ironic logs but the error is from IPA, interesting.

**Comment by dtantsur** [comment]
Ramdisk log collection seems broken in that run :(

```
sed: -e expression #1, char 266: unknown option to `s'
```

**Comment by dtantsur** [comment]
@lentzi90 I think I tracked this down to a bug in IPA: https://review.opendev.org/c/openstack/ironic-python-agent/+/942690

**Comment by lentzi90** [comment]
Thank you @dtantsur ! This has been driving me crazy!

**Comment by Rozzii** [comment]
Thank you @dtantsur !

**Comment by lentzi90** [comment]
I think I found another issue. 😦 

While Ironic was pinned, we [switched to kind](https://github.com/metal3-io/baremetal-operator/pull/2209) because minikube was causing flakes. That obviously means we have not tested this with the latest ironic-image and it seems it does not work.

The difference is how Ironic is exposed to the BMHs. With minikube, we were running everything in the same network, but with kind, we have Ironic in a separate network. This means that we rely on the [IRONIC_EXTERNAL_CALLBACK_URL](https://github.com/metal3-io/baremetal-operator/blob/897ce61e9d7c74b6d5403d6cdc9080a71d8e009b/ironic-deployment/overlays/e2e/ironic_bmo_configmap.env#L4) for the callback and something about this has changed.
I think the issue is that we pass the [base URL here](https://github.com/metal3-io/ironic-image/blob/3a90347170e38d38cb3d9e8b1cfadb7d81a3d489/scripts/runhttpd#L33). That would make IPA use the internal address instead of the external. I will try to test and fix this today.
